pyActigraphy/io/cwa/cwa_class.py

```diff
@@ -1,237 +1,513 @@
+            raise TypeError("filename must be a string")
+        self.__stream = open(filename, "rb")
+        self.__endianess = "<"
+        self.__meta = dict()
+        self.__device = 0
+        self.__name = 0
+        self.__uuid = 0
+        self.__startTime = datetime.min
+        self.__duration = timedelta()
+        self.__period = 0
+        self.__Accel = pandas.DataFrame(columns=["x","y", "z","norm"])
+        self.__Gyro = pandas.DataFrame(columns=["x","y", "z","norm"])
+        self.__Mag = pandas.DataFrame(columns=["x","y", "z","norm"])
+        self.__Ascale = 0
+        self.__Gscale = 0
+        self.__Mscale = 0
+        self.__Others = None
+
+        self.__blocks = 0
+        self.__stream.seek(0,2)
+        self.__blocks = self.__stream.tell() - 1024
+        self.__stream.seek(0)
+        if self.__blocks%512 != 0:
+            raise ValueError("Data file corrupted")
+        self.__blocks = self.__blocks//512
+
+        self._ReadHeader()
+        self.printHeader()
+        
+        t = datetime.now()
+        for blk in range(0, self.__blocks):
+            rem = blk % 1000
+            if rem == 0:
+                print("Block {} of {} ({}%) {}"
+                      .format(blk,self.__blocks,
+                              100.*blk/self.__blocks,
+                              datetime.now() - t)
+                      )
+                t = datetime.now()
+                data = self.__stream.read(512*1000)
+            res = self._ReadBlock(data[512 * rem: 512 * rem + 512])
+            if res is None:
+                break
+            if res[4] is not None:
+                self.__Accel = self.__Accel.append(res[4], ignore_index=False)
+            if res[5] is not None:
+                self.__Gyro = self.__Gyro.append(res[5], ignore_index=False)
+            if res[6] is not None:
+                self.__Mag = self.__Mag.append(res[6], ignore_index=False)
+        print()
+        print("Accelerometer table:")
+        self.__Accel.info()
+        print("Gyrometer table:")
+        self.__Gyro.info()
+        print("Magnetometer table:")
+        self.__Mag.info()
+
+    def _ReadHeader(self, offset=None):
+        """
+        reads and parce the cwa header block
+
+        Parameters
+        ----------
+        offset: int, optional
+            if given, will read from given position
+            if ommited, will read from current position
+        """
+        if offset is not None:
+            self.__stream.seek(offset)
+
+        dr = DataReader(self.__stream.read(1024), self.__endianess)
+
+        # @ 0  +2   ASCII "MD", little-endian (0x444D)
+        tag = dr.unpack("", 2).decode("ASCII")
+        if tag != "MD":
+            self.__stream.close()
+            raise ValueError("{} not a CWA file"
+                             .format(self.__stream.__name__))
+
+        # @ 2  +2   Packet length (1020 bytes, 
+        # with header (4) = 1024 bytes total)
+        length = dr.unpack("uint16")
```

Comment:
@nbeliy : in your doc header of the unpack function, you write that providing the offset reads the data without moving the internal pointer. How comes then you manage to unpack the packet length at position @2 without moving the pointer, since the precedent unpacking does use the offset.

More generally, I think that (unless it harms the reading performance) using the offset explicitly in the unpack function makes the code more readable and in-line with the in-line doc of the CWA header structure (which gives the offset and the length of each attribute).

Reply:
There no strict pointer -- unpack reads data in memory at given index. I simulate the pointer by offset variable, and I decide to move it or not.

Using offset explicitely is error prone: you must track exactly where to read. And in case of packed values an error of one bit not always easely detectable.
That's why I hide offset, as data are readed sequentually. If you need skip some bits, you can always use >>

Reply:
Please re-read the first paragraph of my comment and your doc for your `unpack` function.

For the second paragraph, I meant "using the offset" in the original unpack function (https://docs.python.org/3/library/struct.html#struct.Struct.unpack_from), not yours.